buildSrc/src/main/java/org/elasticsearch/gradle/release/ReleaseNotesGenerator.java

```diff
@@ -0,0 +1,150 @@
+
+        final List<Version> changelogVersions = new ArrayList<>(changelogsByVersion.keySet());
+        changelogVersions.sort(Version::compareTo);
+        Collections.reverse(changelogVersions);
+
+        for (Version version : changelogVersions) {
+            generateForVersion(version, changelogsByVersion.get(version));
+        }
+    }
+
+    private void generateForVersion(Version version, List<ChangelogEntry> changelogs) {
+        Map<String, List<ChangelogEntry>> groupedChangelogs = changelogs.stream()
+            .collect(
+                Collectors.groupingBy(
+                    // Breaking changes come first in the output
+                    entry -> entry.getBreaking() == null ? entry.getType() : "breaking",
+                    TreeMap::new,
+                    Collectors.toList()
+                )
+            );
+
+        generateHeader(version);
+
+        groupedChangelogs.forEach((type, changelogsByType) -> {
+            generateGroupHeader(version, type);
+
+            final TreeMap<String, List<ChangelogEntry>> changelogsByArea = changelogsByType.stream()
+                .collect(Collectors.groupingBy(ChangelogEntry::getArea, TreeMap::new, Collectors.toList()));
+
+            changelogsByArea.forEach((area, perAreaChangelogs) -> {
+                out.println(area + "::");
+
+                // Generate the output lines first so that we can sort them
+                perAreaChangelogs.stream().map(log -> {
+                    final StringBuilder sb = new StringBuilder(
+                        "* " + log.getSummary() + " {es-pull}" + log.getPr() + "[#" + log.getPr() + "]"
+                    );
+                    final List<Integer> issues = log.getIssues();
+                    if (issues != null && issues.isEmpty() == false) {
+                        sb.append(issues.size() == 1 ? " (issue: " : " (issues: ");
+                        sb.append(issues.stream().map(i -> "{es-issue}" + i + "[#" + i + "]").collect(Collectors.joining(", ")));
+                        sb.append(")");
+                    }
+                    return sb.toString();
+                }).sorted().forEach(out::println);
+
+                out.println();
+            });
+
+            out.println();
+        });
+    }
+
+    private void generateHeader(Version version) {
+        String branch = version.getMajor() + "." + version.getMinor();
+
+        out.println("[[release-notes-" + version + "]]");
+        out.println("== {es} version " + version);
+
+        if (version.getLabels().contains("SNAPSHOT")) {
+            out.println();
+            out.println("coming[" + version + "]");
+        }
+
+        out.println();
+        out.println("Also see <<breaking-changes-" + branch + ",Breaking changes in " + branch + ">>.");
+        out.println();
+    }
+
+    private void generateGroupHeader(Version version, String type) {
+        out.println("[[" + type + "-" + version + "]]");
+        out.println("[float]");
```

Comment:
Style nit: We try to use `[discrete]` now per Asciidoctor preferences: https://asciidoctor.org/docs/asciidoc-asciidoctor-diffs/#blocks

Shouldn't make a different to the actual output.
```suggestion
        out.println("[discrete]");
```